media-sound/audacity: fix libsamplerate/libsox dep

marcecj opened the pull request:
Starting with audacity 2.0.3, only *one* resampling library can be enabled at a
time.  Due to this, the way the USE flags are currently handled make it
impossible to use the libsoxr library.  This commit does two things to remedy
this:

- never enable libresample (which isn't recommended anymore), and
- make the libsamplerate and libsoxr USE flags exclusive, reflecting the way
  the build system works.

These two minimal changes make it possible to properly select between the two
recommended resampling libraries.

Gentoo-bug: 536208
Signed-off-by: Marc Joliet <marcec@gmx.de>

[I'm not *entirely* sure if this PR is appropriate.  Audacity-2.1.1 seems to only use libsoxr now, but it's been hard masked since July and it might be good to also have a fixed version of 2.0.5.]

mgorny commented:
Friendly diff:

```
--- audacity-2.0.5.ebuild	2015-08-09 10:19:52.151071161 +0200
+++ audacity-2.0.5-r1.ebuild	2015-10-26 23:56:05.012839085 +0100
@@ -15,7 +15,7 @@ SRC_URI="mirror://gentoo/${MY_T}.tar.xz"
 
 LICENSE="GPL-2"
 SLOT="0"
-KEYWORDS="amd64 ~mips ppc ppc64 x86"
+KEYWORDS="~amd64 ~mips ~ppc ~ppc64 ~x86"
 IUSE="alsa ffmpeg flac id3tag jack ladspa libsamplerate +libsoxr midi mp3 sbsms soundtouch twolame vamp vorbis"
 RESTRICT="test"
 
@@ -47,7 +47,8 @@ DEPEND="${COMMON_DEPEND}
 	app-arch/xz-utils
 	virtual/pkgconfig"
 
-REQUIRED_USE="soundtouch? ( midi )"
+REQUIRED_USE="soundtouch? ( midi )
+	^^ ( libsoxr libsamplerate )"
 
 S=${WORKDIR}/${MY_P}
 
@@ -72,7 +73,6 @@ src_configure() {
 		--with-libsndfile=system \
 		--with-expat=system \
 		$(use_with libsamplerate) \
-		$(use_with !libsamplerate libresample) \
 		$(use_with libsoxr) \
 		$(use_with vorbis libvorbis) \
 		$(use_with mp3 libmad) \

mgorny commented:
@gentoo/proaudio

marcecj commented:
OK, I updated the PR following @aballier's comments (which seem to not have been made via this PR, but can be found at https://github.com/marcecj/gentoo/commit/793fdf25706fdfd51b93963a8a4a3748733d575d instead).  I also replaced a comment I overlooked before.

The new diff is:

    diff --git a/media-sound/audacity/audacity-2.0.5.ebuild b/media-sound/audacity/audacity-2.0.5-r1.ebuild
    similarity index 87%
    copy from media-sound/audacity/audacity-2.0.5.ebuild
    copy to media-sound/audacity/audacity-2.0.5-r1.ebuild
    index 11a5e3d..87c2a09 100644
    --- a/media-sound/audacity/audacity-2.0.5.ebuild
    +++ b/media-sound/audacity/audacity-2.0.5-r1.ebuild
    @@ -15,15 +15,15 @@ SRC_URI="mirror://gentoo/${MY_T}.tar.xz"
     
     LICENSE="GPL-2"
     SLOT="0"
    -KEYWORDS="amd64 ~mips ppc ppc64 x86"
    -IUSE="alsa ffmpeg flac id3tag jack ladspa libsamplerate +libsoxr midi mp3 sbsms soundtouch twolame vamp vorbis"
    +KEYWORDS="~amd64 ~mips ~ppc ~ppc64 ~x86"
    +IUSE="alsa ffmpeg flac id3tag jack ladspa +libsoxr midi mp3 sbsms soundtouch twolame vamp vorbis"
     RESTRICT="test"
     
     COMMON_DEPEND="x11-libs/wxGTK:2.8[X]
            >=app-arch/zip-2.3
            >=media-libs/libsndfile-1.0.0
            dev-libs/expat
    -       libsamplerate? ( >=media-libs/libsamplerate-0.1.2 )
    +       !libsoxr? ( >=media-libs/libsamplerate-0.1.2 )
            libsoxr? ( media-libs/soxr )
            vorbis? ( >=media-libs/libvorbis-1.0 )
            mp3? ( >=media-libs/libmad-0.14.2b )
    @@ -63,7 +63,7 @@ src_configure() {
     
            # * always use system libraries if possible
            # * options listed in the order that configure --help lists them
    -       # * if libsamplerate not requested, use libresample instead.
    +       # * if libsoxr is not requested, use libsamplerate instead
            econf \
                    --enable-unicode \
                    --enable-nyquist \
    @@ -71,8 +71,7 @@ src_configure() {
                    $(use_enable ladspa) \
                    --with-libsndfile=system \
                    --with-expat=system \
    -               $(use_with libsamplerate) \
    -               $(use_with !libsamplerate libresample) \
    +               $(use_with !libsoxr libsamplerate) \
                    $(use_with libsoxr) \
                    $(use_with vorbis libvorbis) \
                    $(use_with mp3 libmad) \

aballier commented:
good, but considering we now have only two choices and libsoxr is the default on later releases, it'd be better to have a libsamplerate useflag, replacing libsoxr by libsamplerate, instead of the contrary

also, it'd be nice to have a local description of libsamplerate useflag, like:
"Use libsamplerate resampler: Much slower than libsoxr one but better quality."

marcecj commented:
I think I see what you mean: it is better to have a USE flag reflect a conscious choice, i.e., USE="libsamplerate" is better than USE="-libsoxr".  I'll push that change after posting this.

Regarding the description: that does not reflect what I read about libsoxr.  It's supposed to be both faster *and* (slightly) higher quality than libsamplerate (see for example [this comparison](http://src.infinitewave.ca/) -- libsoxr and libsamplerate are named "SoX" and "Secret Rabbit Code", respectively).  The only downside I've read of is that it has an increased processing latency relative to libsamplerate (since it's FFT based), making it less suitable for real-time tasks.

So unless I've missed something, the libsoxr USE flag description also needs updating ;-) .

aballier commented:
> I think I see what you mean: it is better to have a USE flag reflect a conscious choice, i.e., USE="libsamplerate" is better than USE="-libsoxr". I'll push that change after posting this.

yes, exactly :)

as for src vs soxr, you're probably right, the description and my understanding of it is from several years ago, when libsoxr was just released, it is very likely it has improved since then

aballier commented:
merged, thanks!

please open a new PR for the useflag description if you wish to update it, as yours is probably more up to date

marcecj commented:
Gladly!  I can look into the USE flag description later today.
